[mem.res.pool.overview] add proper definitions of terms

jwakely opened the pull request:
Update cross-reference in [mem.res.pool.options] to refer to
[mem.res.pool.overview].

Fixes #1258

CaseyCarter commented on source/utilities.tex:
```diff
@@ -11806,30 +11806,54 @@
 Each call to \tcode{do_allocate(size, alignment)} is dispatched
-to the pool serving the smallest blocks accommodating at least \tcode{size} bytes.
+to the pool serving the smallest blocks accommodating at \tcode{size} bytes
+and are aligned to a multiple of \tcode{alignment}.
```
This sentence has gone astray: "... accommodating at `size` bytes and are aligned to a multiple of `alignment`." is not Englishing goodly. How about "Each call to `do_allocate(size, alignment)` is dispatched to the pool serving the smallest blocks that accommodate `size` and `alignment`."? "multiple of `alignment` seems extraneous to me given that all alignments are powers of two (http://eel.is/c++draft/basic.align#4).

jwakely replied:
Yes I'm not sure how that got mangled. I'll push an update.

Quuxplusone commented on source/utilities.tex:
```diff
@@ -11824,30 +11824,54 @@
+with different block sizes.
+
+\item
+On destruction, a pool resource frees all the memory it obtained from
```
"On any call to `release()`..."? And then just specify that the destructor always calls `release()`?
Otherwise we'll have either that `release()` is underspecified, or the specification of `release()` duplicates all this excellent wording.

Quuxplusone commented on source/utilities.tex:
```diff
@@ -11824,30 +11824,54 @@
 allocating a block from that pool results in the allocation
-of an additional chunk of memory from the \defn{upstream allocator}
-(supplied at construction), thus replenishing the pool.
+of an additional chunk from the upstream memory resource,
```
Does "an" imply that only one such chunk must be allocated? Could it be "one or more" chunks?

CaseyCarter replied:
Replacing "an additional chunk" with "additional chunks" would IMO be a simpler way to express that behavior than "one or more additional chunks".

Quuxplusone commented on source/utilities.tex:
```diff
@@ -11824,30 +11824,54 @@
-Allocation requests that exceed the largest block size of any pool
-are fulfilled directly from the upstream allocator.
+Allocation requests that exceed the largest block size and/or alignment
+of any pool are fulfilled directly from the upstream allocator.
```
Preexisting grammar nit: presumably "any pool" should read "all pools"

Quuxplusone commented on source/utilities.tex:
```diff
@@ -11943,7 +11967,7 @@
 \pnum
 The maximum number of blocks that will be allocated at once
-from the upstream memory resource\iref{mem.res.monotonic.buffer}
+from the upstream memory resource\iref{mem.res.pool.overview}
```
If you take my suggestion that the resource be permitted to allocate "one or more" chunks to replenish a pool, then this wording must be updated to replace the phrase "at once" with the phrase "in any single chunk".

jwakely committed:
[mem.res.pool.overview] add proper definitions of terms

Update cross-reference in [mem.res.pool.options] to refer to
[mem.res.pool.overview].

Fixes #1258

jwakely commented:
Rebased, all comments from @CaseyCarter and @Quuxplusone addressed.

jensmaurer commented:
@jwakely , this is a large-ish change, and @zygoloid added the "lwg" label four years ago. Do you think this should go via LWG, or should be just collect some endorsement here from @CaseyCarter and others?

jwakely commented:
I think it should go via LWG. I'll ask for some reflector reviews of this and #2550 and some others with the lwg label.

Quuxplusone commented on source/memory.tex:
```diff
@@ -5802,31 +5802,55 @@
+
+\item
+On any call to \tcode{release()}, a pool resource frees all the memory
+it obtained from its upstream memory resource by calling
+the upstream resource's \tcode{deallocate} function.
```
`it has obtained`
Abbreviating as `the upstream resource's` here feels just a little weird to me, since the true name is `the upstream memory resource's` and if we're aiming for brevity we could just say `the upstream's`.

I admit that my eye caught on this paragraph originally because it refers to "the memory
it [has] obtained from its upstream memory resource by calling the upstream resource's deallocate function"; when in fact that memory was obtained by calling the upstream's _allocate_ function! But I have no suggested fix for that grammar quirk, so my brain found other nits.

jwakely replied:
How about:

> On a call to `release()`, a pool resource calls its upstream memory resource's `deallocate` function to return all the memory that was obtained from its upstream memory resource.

jwakely replied:
Or "that was obtained from upstream".

Quuxplusone replied:
Minor tweak:

> On a call to `release()`, a pool resource calls its upstream memory resource's `deallocate` function to return all the memory that _has been_ obtained from its upstream memory resource.

Or "all the _outstanding_ memory," but that's a little jargony. Could even say "all the _chunks_ that have been obtained," if that's correct. Can't say "_has ever been_ obtained," because we really do mean only the outstanding chunks, not ones that were allocated before the last call to `release()` nor ones that have already been deallocated via `deallocate`.

Wildly ambitious rewrite:

> On a call to `release()`, a pool resource calls its upstream memory resource's `deallocate` function on every chunk that has not yet been deallocated.

jwakely replied:
But it doesn't say all memory that has *ever* been allocated, and we don't bother making that distinction for containers freeing memory with an allocator.

Quuxplusone commented on source/memory.tex:
```diff
@@ -5802,31 +5802,55 @@
-Allocation requests that exceed the largest block size of any pool
-are fulfilled directly from the upstream allocator.
+Allocation requests that exceed the largest block size and/or alignment
+of all pools are fulfilled directly from the upstream allocator.
```
`from the upstream memory resource`?

Quuxplusone reviewed (commented):
Some nits, but LGTM FWIW!

JohelEGP commented on source/memory.tex:
```diff
@@ -5802,31 +5802,55 @@
+\item
+A \defn{chunk} is a region of storage allocated
+from the upstream memory resource. A pool resource divides
+each chunk into one or more blocks of the same block size.
```
No. `\defn` is for marking the definition, not its uses.

jwakely committed:
FIXUP review comments
